### ISSUE TITLE
.chezmoi.username on Android is wrong

### ISSUE THREAD

**Issue by ErrrorMaxx** [twpayne/chezmoi#3708]
## Describe the bug

For a long time I had line like `export USER=$(whoami) LOGNAME=$(whoami)` in my .zshrc file, so I always had `$USER` and `$LOGNAME` set to a correct value

Today I decided to delete this line and got strange behaviour of `chezmoi`: it now reports `.chezmoi.username` as `"android"` instead of `"u0_aXXX"`

After digging into source code I found this line: https://github.com/twpayne/chezmoi/blob/ad74250e8b2fbc063080c907e06e5f9b927c3d8b/internal/cmd/config.go#L2260

I checked this and calling `chezmoi` with different `$USER` actually changes `.chezmoi.username`!

## To reproduce

```shell
# Real username
$ whoami
u0_a324

# Incorrect
$ chezmoi execute-template '{{ .chezmoi.username }}'
android

# What I found
$ USER=wtf chezmoi execute-template '{{ .chezmoi.username }}'
wtf
```

## Expected behavior

`.chezmoi.username` is correct on Android

## Additional information

I think that checking `$LOGNAME` first and checking `$USER` if it failed is better fallback for username detection, because:
- `$LOGNAME` is from POSIX standard, it is set by Termux and also used/set by many other tools/apps
- `$USER` is from BSD and is non-standard and is not guaranteed to exist

## Output of `chezmoi doctor`

<details>

```console
$ chezmoi doctor
RESULT    CHECK                       MESSAGE
warning   version                     v2.47.4
ok        latest-version              v2.47.4
ok        os-arch                     android/arm64
ok        uname                       Linux localhost 4.19.191+ #1 SMP PREEMPT Tue Mar 26 11:53:41 CST 2024 aarch64 Android
ok        go-version                  go1.22.2 (gc)
ok        executable                  /data/data/com.termux/files/usr/bin/chezmoi
ok        config-file                 ~/.config/chezmoi/chezmoi.toml, last modified 2024-04-20T18:31:36Z
warning   source-dir                  ~/.local/share/chezmoi is a git working tree (dirty)
warning   suspicious-entries          ~/.local/share/chezmoi/home/.chezmoiexternal.yml.tmpl and ~/.local/share/chezmoi/home/dot_config/rootmoi/chezmoi.yaml.tmpl
warning   working-tree                ~/.local/share/chezmoi is a git working tree (dirty)
ok        dest-dir                    ~ is a directory
warning   umask                       077
ok        cd-command                  found /data/data/com.termux/files/usr/bin/zsh
ok        cd-args                     /data/data/com.termux/files/usr/bin/zsh
info      diff-command                not set
ok        edit-command                found ~/.local/bin/u-editor
ok        edit-args                   ~/.local/bin/u-editor
ok        git-command                 found /data/data/com.termux/files/usr/bin/git, version 2.44.0
warning   merge-command               vimdiff not found in $PATH
ok        shell-command               found /data/data/com.termux/files/usr/bin/zsh
ok        shell-args                  /data/data/com.termux/files/usr/bin/zsh
info      age-command                 age not found in $PATH
ok        gpg-command                 found /data/data/com.termux/files/usr/bin/gpg, version 2.4.4
info      pinentry-command            not set
info      1password-command           op not found in $PATH
info      bitwarden-command           bw not found in $PATH
info      bitwarden-secrets-command   bws not found in $PATH
info      dashlane-command            dcli not found in $PATH
info      doppler-command             doppler not found in $PATH
info      gopass-command              gopass not found in $PATH
info      keepassxc-command           keepassxc-cli not found in $PATH
info      keepassxc-db                not set
info      keeper-command              keeper not found in $PATH
info      lastpass-command            lpass not found in $PATH
info      pass-command                pass not found in $PATH
info      passhole-command            ph not found in $PATH
info      rbw-command                 rbw not found in $PATH
info      vault-command               vault not found in $PATH
info      vlt-command                 vlt not found in $PATH
info      secret-command              not set
```

</details>

**Comment by halostatue** [comment]
TL;DR: I disagree that this is the order that should be used, and am not sure that this is a desirable change. *Maybe* if it is narrowed to android/termux only.

> I think that checking `$LOGNAME` first and checking `$USER` if it failed is better fallback for username detection, because:
> 
> * `$LOGNAME` is from POSIX standard, it is set by Termux and also used/set by many other tools/apps
> * `$USER` is from BSD and is non-standard and is not guaranteed to exist

This change will break in some circumstances. Whether they are circumstances that chezmoi should be supporting is a valid question, but the change suggested cannot be adopted without discussion on this.

I personally think that this is a Termux bug (it should not be setting `$LOGNAME` but should be setting `$USER`), because `$LOGNAME` is supposed to be the logged-in user's name, whereas `$USER` is supposed to be the current interactive user's name. 

Depending on the system you’re using, `sudo bash -c 'echo $LOGNAME $USER'` should print *different* words (on macOS, they print the same value) and use in something like a cron job is likely to be different (although I suspect there `$LOGNAME` will be set, but `$USER` would be unset).

`$USER` is from BSD, but something which has existed and used for > 40 years is fair to call a _de facto_, but not necessarily _de jure_ standard.

**Comment by ErrrorMaxx** [comment]
In some (rare) cases on some operating systems I tested there sometimes were no `$USER`, but was `$LOGNAME`, or _when `$LOGNAME` was updated and `$USER` were not_, but most of the time they are actually the same

> because $LOGNAME is supposed to be the logged-in user's name, whereas $USER is supposed to be the current interactive user's name.

They are actually used for the same purpose and both are supported by programs for compatibility reasons

> I personally think that this is a Termux bug (it should not be setting $LOGNAME but should be setting $USER), because $LOGNAME is supposed to be the logged-in user's name, whereas $USER is supposed to be the current interactive user's name.

In my opinion `$LOGNAME` should be prioritized, and `$USER` is used as fallback for compatibility reasons, not vice versa. But I agree that Termux should set `$USER` for compatibility

**Comment by twpayne** [comment]
Thanks for the discussion here.

Firstly, if you want to use `$LOGNAME` in your templates then you can always use `{{ env "LOGNAME" }}` instead of `{{ .chezmoi.username }}`.

If you want to avoid typing `{{ env "LOGNAME" }}` everywhere then you can add

```
[data]
    logname = {{ env "LOGNAME" | quote }}
```

to your config file template and use `{{ .logname }}` instead.

Secondly, chezmoi's username detection is significantly more complex than `os.LookupEnv("USER")`. See: https://github.com/twpayne/chezmoi/blob/4a9d0e4d5d04f7812678607151fd20cfea843e37/internal/cmd/config.go#L2223-L2267

Specifically, `$USER` is only used if Go's standard library function [`os/user.Current()`](https://pkg.go.dev/os/user#Current) fails.

Thirdly, when you run chezmoi, you generally want it to run as your current user, not the user that you logged in as. For example, if you use `sudo` or `su` you generally want to manage the current user's dotfiles, not the dotfiles of the user that you logged in as.

Fourthly, it's worth mentioning that Android is quite different from Linux, even though their kernels share a huge amount of code. These differences, particularly in userspace, are visible in Termux.

Therefore, I propose to change chezmoi so that, on Android (Termux) only, chezmoi prioritizes `$LOGNAME` over `$USER`. For all other systems, chezmoi's behavior remains unchanged.

**Comment by twpayne** [comment]
Fixed with #3712.
